grpc/runtime/src/main/scala/io/buoyant/grpc/runtime/DecodingStream.scala

```diff
@@ -0,0 +1,288 @@
+   * being constructed, it is paired with a _releasable_ function that
+   * frees the underlying H2 frames as needed.
+   */
+  private trait Releaser {
+
+    /**
+     * Add the given H2 frame to be tracked and released.
+     */
+    def track(frame: h2.Frame): Releaser
+
+    /**
+     * Mark `n` bytes as consumed.
+     */
+    def consume(n: Int): Releaser
+
+    /**
+     * Reap all consumed frames into a releasable function, and return
+     * the updated state afterwards.
+     */
+    def releasable(): (Releaser, Releaser.Func)
+  }
+
+  private object Releaser {
+    type Func = () => Future[Unit]
+    val NopFunc: Func = () => Future.Unit
+
+    class Underflow extends Exception("released too many bytes") with NoStackTrace
+
+    object Empty extends Releaser {
+      override def track(f: h2.Frame): Releaser = mk(f)
+
+      override def consume(n: Int): Releaser =
+        if (n == 0) this
+        else throw new Underflow
+
+      override def releasable(): (Releaser, Func) = (this, NopFunc)
+    }
+
+    private[this] case class FrameReleaser(
+      remaining: Int,
+      release: () => Future[Unit],
+      next: Releaser
+    ) extends Releaser {
+      require(remaining >= 0)
+
+      override def track(f: h2.Frame): Releaser =
+        copy(next = next.track(f))
+
+      override def consume(n: Int): Releaser =
+        if (n == 0) this
+        else if (n >= remaining) {
+          // Leave a 0-length releasable, so that `release` is chained
+          // into the next `releasable` call.
+          copy(
+            remaining = 0,
+            next = next.consume(n - remaining)
+          )
+        } else copy(remaining = remaining - n)
+
+      override def releasable(): (Releaser, Func) =
+        if (remaining == 0) {
+          val (rest, releaseRest) = next.releasable()
+          val doRelease = () => release().before(releaseRest())
+          rest -> doRelease
+        } else {
+          // This frame isn't yet releasable, but we require that this
+          // sub-slice is released before releasing the frame:
+          val p = new Promise[Unit]
+          val rest = copy(release = () => p.before(release()))
+          val doRelease = () => { p.setDone(); Future.Unit }
+          rest -> doRelease
+        }
+    }
+
+    private[this] def mk(frame: h2.Frame): Releaser = frame match {
+      case f: h2.Frame.Data => FrameReleaser(f.buf.length, f.release, Empty)
+      case f: h2.Frame.Trailers => FrameReleaser(0, f.release, Empty)
+    }
+  }
+
+  private[this] case class Header(compressed: Boolean, size: Int)
+
+  private[this] val GrpcFrameHeaderSz = Codec.GrpcFrameHeaderSz
+
+  private val hasHeader: ByteBuffer => Boolean =
+    bb => { GrpcFrameHeaderSz <= bb.remaining }
+
+  private val hasMessage: (ByteBuffer, Int) => Boolean =
+    (bb, msgsz) => { msgsz <= bb.remaining }
+
+  private def decodeHeader(bb0: ByteBuffer): Header = {
+    require(hasHeader(bb0))
+    val bb = bb0.duplicate()
```

Comment:
AIUI, the main difference with slice is that it resets the start of the buffer so that `bb.position == 0` and that it sets the _capacity_ to the _limit_ of the buffer.  I don't think buys us much but I'll take a look.